util/teststorage/appender.go

```diff
@@ -0,0 +1,415 @@
+	for _, s := range samples {
+		ret = append(ret, sample{Sample: s})
+	}
+	return ret
+}
+
+// ToSeriesSet converts []Samples to storage.SeriesSet.
+func ToSeriesSet(samples []Sample) storage.SeriesSet {
+	var (
+		series        []storage.Series
+		seriesSamples []Sample
+	)
+	lastL := labels.Labels{}
+	for _, s := range samples {
+		if !labels.Equal(s.L, lastL) {
+			if len(seriesSamples) > 0 {
+				series = append(series, storage.NewListSeries(lastL, ToChunkSampleSlice(seriesSamples)))
+				seriesSamples = seriesSamples[:0]
+			}
+			lastL = s.L
+		}
+		seriesSamples = append(seriesSamples, s)
+	}
+	if len(seriesSamples) > 0 {
+		series = append(series, storage.NewListSeries(lastL, ToChunkSampleSlice(seriesSamples)))
+	}
+	return storage.NewListSeriesSet(series)
+}
+
+// Appendable is a storage.Appendable mock.
+// It allows recording all samples that were added through the appender and injecting errors.
+type Appendable struct {
+	appendErrFn          func(ls labels.Labels) error // If non-nil, inject appender error on every Append, AppendHistogram and ST zero calls.
+	appendExemplarsError error                        // If non-nil, inject exemplar error.
+	commitErr            error                        // If non-nil, inject commit error.
+
+	mtx sync.Mutex // mutex for result writes and ResultSamplesGreaterThan read.
+
+	// Recorded results.
+	pendingSamples    []Sample
+	resultSamples     []Sample
+	rolledbackSamples []Sample
+
+	// Optional chain (Appender will collect samples, then run next).
+	next storage.Appendable
+}
+
+// NewAppendable returns mock Appendable.
+func NewAppendable() *Appendable {
+	return &Appendable{}
+}
+
+// Then chains another appender from the provided appendable for the Appender calls.
+func (a *Appendable) Then(appendable storage.Appendable) *Appendable {
+	a.next = appendable
+	return a
+}
+
+// WithErrs allows injecting errors to the appender.
+func (a *Appendable) WithErrs(appendErrFn func(ls labels.Labels) error, appendExemplarsError, commitErr error) *Appendable {
+	a.appendErrFn = appendErrFn
+	a.appendExemplarsError = appendExemplarsError
+	a.commitErr = commitErr
+	return a
+}
+
+// PendingSamples returns pending samples (samples appended without commit).
+func (a *Appendable) PendingSamples() []Sample {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+
+	ret := make([]Sample, len(a.pendingSamples))
+	copy(ret, a.pendingSamples)
+	return ret
+}
+
+// ResultSamples returns committed samples.
+func (a *Appendable) ResultSamples() []Sample {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+
+	ret := make([]Sample, len(a.resultSamples))
+	copy(ret, a.resultSamples)
+	return ret
+}
+
+// RolledbackSamples returns rolled back samples.
+func (a *Appendable) RolledbackSamples() []Sample {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+
+	ret := make([]Sample, len(a.rolledbackSamples))
+	copy(ret, a.rolledbackSamples)
+	return ret
+}
+
+func (a *Appendable) ResultReset() {
+	a.pendingSamples = a.pendingSamples[:0]
```

Comment:
Shouldn't you take the mutex here, when you're changing these fields, same as in `Append()` for example?
```suggestion
	a.mtx.Lock()
	defer a.mtx.Unlock()
```